src/Raven.Server/Monitoring/ServerLimitsMonitor.cs

```diff
@@ -0,0 +1,214 @@
+            {
+                throw new InvalidOperationException($"Executing '{proc.StartInfo.FileName} {proc.StartInfo.Arguments}' failed (exit code {proc.ExitCode}){Environment.NewLine}StandardOutput:{Environment.NewLine}{proc.StandardOutput.ReadToEnd()}{Environment.NewLine}StandardError:{Environment.NewLine}{proc.StandardError.ReadToEnd()}");
+            }
+
+            var currentPidMax = GetEolCount(proc.StandardOutput.BaseStream);
+
+            AddAlertIfNeeded(currentPidMax, maxPidMax, MaxPidMaxThreshold, pid_max, nameof(pid_max));
+        }
+
+        private void CheckThreadsMaxLimits()
+        {
+            var maxThreadsMax = GetMaxValueForLimit(threads_max);
+
+            var currentThreadsMax = 0L;
+            foreach (var _ in Directory.EnumerateDirectories("/proc/self/task", "*", SearchOption.TopDirectoryOnly))
+            {
+                currentThreadsMax++;
+            }
+
+            AddAlertIfNeeded(currentThreadsMax, maxThreadsMax, MaxThreadsMaxThreshold, threads_max, nameof(threads_max));
+        }
+
+        private void CheckMaxMapCountLimits()
+        {
+            var maxMapCount = GetMaxValueForLimit(max_map_count);
+
+            long currentMapCount = GetCurrentMapCount();
+            AddAlertIfNeeded(currentMapCount, maxMapCount, MaxMapCountThreshold, max_map_count, nameof(max_map_count));
+        }
+
+        private static long GetCurrentMapCount()
+        {
+            long currentMapCount;
+            using (FileStream stream = File.OpenRead("/proc/self/maps"))
+            {
+                currentMapCount = GetEolCount(stream);
+            }
+
+            return currentMapCount;
+        }
+
+        private long GetMaxValueForLimit(string limit)
+        {
+            if (_maxLimitsDictionary.TryGetValue(limit, out var maxValueForLimit) == false)
+            {
+                var maxValueString = File.ReadAllText(limit);
+                if (long.TryParse(maxValueString, out var maxValueLong) && maxValueLong > 0)
+                {
+                    _maxLimitsDictionary.Add(limit, maxValueLong);
+                    return maxValueLong;
+                }
+
+                throw new InvalidOperationException($"Could not parse the value of '{limit}', got: '{maxValueString}' and '{maxValueForLimit}'.");
+            }
+
+            return maxValueForLimit;
+        }
+
+        private void AddAlertIfNeeded(long currentMax, long maxMax, float threshold, string limit, string name)
+        {
+            if (currentMax >= (long)(maxMax * (1 - threshold)))
+            {
+                _alerts.Add($"'{name}' is ({currentMax} / {maxMax}), please increase the '{limit}' limit.");
+            }
+        }
+
+        private static long GetEolCount(Stream stream)
+        {
+            long eolCount = 0L;
+            char cr = '\r';
+            char lf = '\n';
+            char NULL = (char)0;
+
+            byte[] byteBuffer = new byte[256 * 1024];
```

Comment:
You are allocating a huge buffer each call, can use `stackalloc` here

Reply:
ok I will use `PipeReader`